src/aspire/abinitio/commonline_d2.py

```diff
@@ -0,0 +1,1888 @@
+from aspire.utils import J_conjugate, Rotation, all_pairs, all_triplets, tqdm, trange
+from aspire.utils.random import randn
+from aspire.volume import DnSymmetryGroup
+
+logger = logging.getLogger(__name__)
+
+
+class CLSymmetryD2(CLOrient3D):
+    """
+    Define a class to estimate 3D orientations using common lines methods for
+    molecules with D2 (dihedral) symmetry.
+
+    Corresponding publication:
+    E. Rosen and Y. Shkolnisky,
+    Common lines ab-initio reconstruction of D2-symmetric molecules,
+    SIAM Journal on Imaging Sciences, volume 13-4, p. 1898-1994, 2020
+    """
+
+    def __init__(
+        self,
+        src,
+        n_rad=None,
+        n_theta=None,
+        max_shift=0.15,
+        shift_step=1,
+        grid_res=1200,
+        inplane_res=5,
+        eq_min_dist=7,
+        epsilon=0.01,
+        seed=None,
+        mask=True,
+    ):
+        """
+        Initialize object for estimating 3D orientations for molecules with D2 symmetry.
+
+        :param src: The source object of 2D denoised or class-averaged images with metadata
+        :param n_rad: The number of points in the radial direction of Fourier image.
+        :param n_theta: The number of points in the theta direction of Fourier image.
+        :param max_shift: Maximum range for shifts as a proportion of resolution. Default = 0.15.
+        :param shift_step: Resolution of shift estimation in pixels. Default = 1 pixel.
+        :param grid_res: Number of sampling points on sphere for projetion directions.
+            These are generated using the Saaf-Kuijlaars algorithm. Default value is 1200.
+        :param inplane_res: The sampling resolution of in-plane rotations for each
+            projetion direction. Default value is 5 degrees.
+        :param eq_min_dist: Width of strip around equator projection directions from
+            which we DO NOT sample directions. Default value is 7 degrees.
+        :param epsilon: Tolerance for J-synchronization power method.
+        :param seed: Optional seed for RNG.
+        :param mask: Option to mask `src.images` with a fuzzy mask (boolean).
+            Default, `True`, applies a mask.
+        """
+
+        super().__init__(
+            src,
+            n_rad=n_rad,
+            n_theta=n_theta,
+            max_shift=max_shift,
+            shift_step=shift_step,
+            mask=mask,
+        )
+
+        self.grid_res = grid_res
+        self.inplane_res = inplane_res
+        self.n_inplane_rots = int(360 / self.inplane_res)
+        self.eq_min_dist = eq_min_dist
+        self.seed = seed
+        self.epsilon = epsilon
+
+        self.triplets = all_triplets(self.n_img)
+        self.pairs, self.pairs_to_linear = all_pairs(self.n_img, return_map=True)
+        self.n_pairs = len(self.pairs)
+
+        # D2 symmetry group.
+        # Rearrange in order Identity, about_x, about_y, about_z.
+        self.gs = DnSymmetryGroup(order=2, dtype=self.dtype).matrices[[0, 3, 2, 1]]
```

Comment:
This ordering was necessary for matching exact matlab outputs. This ordering also corresponds to the logic dealing with x, y, z equators (for instance in `_generate_scl_angles`). This reordering can be removed and the algorithm will still estimate the rotations, but a bunch of the code comments will be inaccurate (ie. referring to a wrong axis) and it looks like one test will need to be altered. 

I'm ok with removing this if you think it would be better, but I'd probably wait and make it one of the last changes since it will change the output of every function.

Another possibility would be to just change how the matrices in the `SymmetryGroup` object are formed. This would not effect the construction of the D2 volume because every symmetry group element is applied to each gaussian blob used to form the volume, the order in which we apply them doesn't matter.

Reply:
Either way we should add a comment that the order is specific (important) to reproducing the MATLAB code.  Probably make more sense to just keep it here for now.  Sounds like changing it would make a lot of treacherous work with little reward.

Reply:
Sounds good. Added a comment. Thanks.